### PR TITLE
feat(frontend): Add Create subscription in frontend

### PR DESCRIPTION
**Pull request by xxhZs**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

In this pr,

We added `subscription` to the `rw_subscriptions` `rw_relation` and `rw_relation_info`.
We support the parse, build_plan and execution of the statements
```
CREATE SUBSCRIPTION s1 FROM aaa WITH (retention = '1D') ;
DROP SUBSCRIPTION xxx;
ALTER SUBSCRIPTION x1 RENAME TO x2 ;
ALTER SUBSCRIPTION x1 OWNER TO user1;
ALTER SUBSCRIPTION x1 SET SCHEMA test_schema;
ALTER SUBSCRIPTION x1 SET PARALLELISM = 4;
SHOW SUBSCRIPTIONS;
SHOW CREATE SUBSCRIPTION subscription1;
```

And in this pr, We have only added tests for SUBSCRIPTION ddl, the specific result tests will be added in the next pr

related #14316

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/handler/mod.rs
Comment (hzxa21) [inline src/frontend/src/handler/mod.rs:804]: Have we tested all the ALTER operations work for for subscription?

File: src/frontend/src/observer/observer_manager.rs
Comment (hzxa21) [inline src/frontend/src/observer/observer_manager.rs:119]: Why ignore `subscriptions` here? Does FE need to cache the subscription catalog? If yes, we should store the `subscriptions` in the FE catalog cache. If not, we should remove `subscriptions` from `MetaSnapshot`

**Review by hzxa21 — COMMENTED** [review]
I am a bit concerned with a 2.3K LoC PR with zero test :(

Is it possible to split it into smaller PRs and add some tests (at least e2e tests).

File: proto/stream_plan.proto
Comment (xxhZs) [reply proto/stream_plan.proto:724]: yesyes

**Comment by xxhZs** [comment]
> I am a bit concerned with a 2.3K LoC PR with zero test :(
> 
> Is it possible to split it into smaller PRs and add some tests (at least e2e tests).

I have added ci, including alter, and simple create, drop statements, (ci for correctness testing will be added with the cursor implementation)

**Comment by wenym1** [comment]
> I am a bit concerned with a 2.3K LoC PR with zero test :(
> 
> Is it possible to split it into smaller PRs and add some tests (at least e2e tests).

+1 for splitting the PR. 

The current PR can be split into at least 2 PRs. One for meta part only, which involves adding new metadata model and new meta node rpc, and one for frontend part that handles the SQL and call the meta node rpc.

**Comment by xxhZs** [comment]
> > I am a bit concerned with a 2.3K LoC PR with zero test :(
> > Is it possible to split it into smaller PRs and add some tests (at least e2e tests).
> 
> +1 for splitting the PR.
> 
> The current PR can be split into at least 2 PRs. One for meta part only, which involves adding new metadata model and new meta node rpc, and one for frontend part that handles the SQL and call the meta node rpc. The PRs can be further split if a single PR has more than 1K LOC.

okk,  I split it 
meta part is in https://github.com/risingwavelabs/risingwave/pull/15371

**Comment by xxhZs** [comment]
After spilit out the meta code. the pr is left with only the frontend code which is slightly more than 1k in addition to the tests, the pr can continue to be reviewed

**Comment by wenym1** [comment]
Left a comment in the previous PR. Might be fixed by the way in this PR.
https://github.com/risingwavelabs/risingwave/pull/15371#discussion_r1512182710

File: src/frontend/src/optimizer/plan_node/stream_sink.rs
Comment (hzxa21) [inline src/frontend/src/optimizer/plan_node/stream_sink.rs:516]: Let's move this method out of `StreamSink` (maybe into `util`) given that it is not just used by sink.

**Review by hzxa21 — APPROVED** [review]
LGTM. Please wait for other reviewers' review who are more familiar with optimizer/parser before merging.

**Review by chenzl25 — COMMENTED** [review]
Could you please test drop cascade for a table if it is used by a subscription?

```
create table t (..);
create subscription sub from t;
drop table t cascade;
```

**Comment by xxhZs** [comment]
> Could you please test drop cascade for a table if it is used by a subscription?
> 
> ```
> create table t (..);
> create subscription sub from t;
> drop table t cascade;
> ```

test it, `drop table t cascade;` will drop `sub`;

**Review by chenzl25 — APPROVED** [review]
LGTM!

**Comment by chenzl25** [comment]
> > Could you please test drop cascade for a table if it is used by a subscription?
> > ```
> > create table t (..);
> > create subscription sub from t;
> > drop table t cascade;
> > ```
> 
> test it, `drop table t cascade;` will drop `sub`;

I see, no wonder I couldn't find the relevant code in this PR, because it is supported in a previous PR.

**Pull request merged**
